File: EscapeRoom/src/gestorAplicacion/clasesLogicasHerencia/Robot.java
package gestorAplicacion.clasesLogicasHerencia;
import gestorAplicacion.clasesLogicas.*;
import java.io.Serializable;
import java.util.ArrayList;
import java.util.Objects;
import uiMain.Main;


public class Robot extends Individuo implements Serializable {
	
	private static final long serialVersionUID= 1L;
	private static ArrayList<Robot> robots = new ArrayList<Robot>();
		public static ArrayList<Robot> getRobots() {
		return robots;
	}

		private static final int ATTACK = 12;
        private boolean aware,nextTo;//atributo que guarda si el robot sabe donde esta el intruso
        private boolean cargaRobot; //para ataque cargado
        private Habitacion goingTo;//hacia donde se va a mover

    public Robot(int health, int armor, int speed, boolean aware, boolean nextTo, boolean cargaRobot) {
			super(health, armor, speed);
			this.aware = aware;
			this.nextTo = nextTo;
			this.cargaRobot = cargaRobot;
			robots.add(this);
		}

	public Robot() {
        this(150, 3, 7, false, false, false);
    }
    
    //Metodos de busqueda
    public void escuchar(Habitacion[] casa){
        if (this.getHealth() > 0){
            for (Habitacion hab : casa) {
                if (hab.getAlarma().equals(Ahorro.ENCENDIDO)) {
                    this.aware = true;
                    goingTo = hab;
                    break;
                } else {
                    this.aware = false;
                }
            }
        }
            
    }
    public void escanear(){
        if (this.getHealth() > 0){
            ArrayList<Habitacion> disponibles = new ArrayList<>();
            disponibles.add(this.getUbicacion().getNorte());
	    disponibles.add(this.getUbicacion().getSur());
	    disponibles.add(this.getUbicacion().getEste());
	    disponibles.add(this.getUbicacion().getOeste());
            for(Habitacion Hab: disponibles) {
                if (!Objects.isNull(Hab) && Hab.getLuces().equals(Ahorro.ENCENDIDO)) {
                    nextTo = true;
                    goingTo = Hab;
                    break;
	        } else {
                    nextTo = false;
                }
	    }
        }
        
    }
    public void apagarAlarma(){
        if (this.getHealth() > 0 && this.getUbicacion().getAlarma().equals(Ahorro.ENCENDIDO)){
            this.getUbicacion().setAlarma(Ahorro.ACTIVADO);
        }
    }
    public Habitacion buscar(Habitacion[] casa){//busqueda en casa 3x3 habitaciones
        Habitacion hab = null;
        if (this.getHealth() > 0){
            ArrayList<Habitacion> disponibles = new ArrayList<>();
	disponibles.add(this.getUbicacion().getNorte());
	disponibles.add(this.getUbicacion().getSur());
	disponibles.add(this.getUbicacion().getEste());
	disponibles.add(this.getUbicacion().getOeste());
        int j = 0;
        for (Habitacion d : disponibles){
            if(Objects.isNull(d)){
                j++;
            }
        }
        if (!Objects.isNull(goingTo.getNorte()) && disponibles.contains(goingTo.getNorte())){
            hab = goingTo.getNorte();
        } else if (!Objects.isNull(goingTo.getSur()) && disponibles.contains(goingTo.getSur())){
            hab = goingTo.getSur();
        } else if (!Objects.isNull(goingTo.getEste()) && disponibles.contains(goingTo.getEste())){
            hab = goingTo.getEste();
        } else if (!Objects.isNull(goingTo.getOeste()) && disponibles.contains(goingTo.getOeste())){
            hab = goingTo.getOeste();
        } else if(j == 2){
            for (Habitacion d : disponibles){
            if(!Objects.isNull(d)){
                hab = d;
            }
            }
        } else {
            hab = casa[4];
        }
        } 
        return hab;
    }
    


    //METODOS PELEAR
    @Override
    public void atacar(Individuo i) {
        i.setHealth(i.getHealth()-ATTACK);
    }
    public String ataqueCargado(Individuo i){
        cargaRobot = false;
        if (Main.lanzarDados(5) >= i.getArmor()) {
            this.atacar(i);
            this.atacar(i);
            this.atacar(i);
            return "El robot lanza un poderoso laser hacia ti!!";
        } else {
            return "El robot lanza un poderoso laser hacia ti! Por suerte, logras esquivarlo";
	}
    }
    public String turno(int desicion,Individuo i){
        if (desicion < 6) {// del 1 al 5 ataque normal
            if (Main.lanzarDados(5) >= i.getArmor()) {
                this.atacar(i);
                return "El robot te acaba de asestar un golpe";
            } else {
                return "Bloqueaste el golpe del robot";
	    }

	} else if (desicion == 6 || desicion == 7) {// 6 o 7 recarga de ataque cargado
            cargaRobot = true;
            return "El pecho del robot comienza a brillar con fuerza";
	} else if (desicion == 8 || desicion == 9) {// 8 o 9 stunear
            if (Main.lanzarDados(5) >= i.getArmor()) {
                i.stun(true);
                return "El robot te electrocutó, estarás aturdido por el siguiente turno";
	    } else {
		return "Bloqueaste el golpe del robot";
	    }
	} else if (desicion == 10) {// 10 roba un objeto
            if (!((Intruso)i).getObjectInventory().isEmpty()){
            int dados = Main.lanzarDados(((Intruso)i).getObjectInventory().size())-1;
            Herramientas quitada = ((Intruso)i).getObjectInventory().get(dados);
            this.getUbicacion().getListaObjetos().add(quitada);
            ((Intruso)i).getObjectInventory().remove(dados);
	    return "El robot logra quitarte " + ((Objetos)quitada).getName() + " y lo arroja al suelo.";
            } else {
                return "El robot intenta quitarte algo... pero aun no has recogido nada.";
            }
            
	} else {
            return "";
        }
    }
    
    //METODOS GET
    public static int getATTACK(){
        return ATTACK;
    }
    public boolean isAware(){
        return aware;
    }
    public boolean isNextTo(){
        return nextTo;
    }
    public boolean isCargaRobot(){
        return cargaRobot;
    }
    public Habitacion getGoingTo(){
        return goingTo;
    }
    
    //METODOS SET
    public void setAware(boolean b){
        aware = b;
    }

    @Override
    public String ayudaJarvis() {
        String a;
        if(aware || nextTo){
            a = "sabe dónde estás.";
        } else {
            a = "no nota tu presencia... aún.";
        }
        if (this.getHealth() > 0){
                    return "El robot se encuentra en la habitacion " + this.getUbicacion().getNumero() + ", tiene " + this.getHealth() + " puntos de vida y " + a;
        } else {
            return "El robot ya ha sido destruido";
        }
    }
    
    @Override
    public void addHistorial() {
        Robot.getHistorial().add("Robot se movio a la habitacion " + this.getUbicacion().getNumero());
    }

    @Override
    public void mover(Habitacion hab) {
        if (this.getHealth() > 0){
            this.getUbicacion().setRobot(null);
            this.setUbicacion(hab);
            this.getUbicacion().setRobot(this);
            this.addHistorial(); 
        } else {
            this.setUbicacion(null);
        }  
    }

   public int decidirDireccion(){
       ArrayList<Habitacion> disponibles = new ArrayList<>();
	disponibles.add(this.getUbicacion().getNorte());
	disponibles.add(this.getUbicacion().getSur());
	disponibles.add(this.getUbicacion().getEste());
	disponibles.add(this.getUbicacion().getOeste());
        ArrayList<Integer> i = new ArrayList<>();
        for(Habitacion Hab: disponibles) {
            if (!Objects.isNull(Hab)) {
                i.add(Hab.getNumero());	    }
	}
        return i.get(Main.lanzarDados(i.size())-1);
   }
   
        @Override
    public String toString(){
        
        String m = "El Robot tiene... vida: " + this.getHealth() + " armadura: " + this.getArmor() + " velocidad: " + this.getSpeed() + " ataque: " + ATTACK;
        m +=  ", esta en la habitacion " + this.getUbicacion().getNumero();
        return m;
    }
}

File: python/Robot.py
import math
import random
from individuo import Individuo
from ahorro import Ahorro
#from main import Main

class Robot(Individuo):
    
    @staticmethod
    def lanzarDados(caras):
        return int(math.floor(random.random() * (caras) + 1))


    _robots = []
    @staticmethod
    def getRobots():
        return Robot._robots

    _ATTACK = 12

    def __init__(self, health = 150, armor = 3, speed = 7, aware = False, nextTo = False, cargaRobot = False):
        self._goingTo = None

        super().__init__(health, armor, speed)
        self._aware = aware
        self._nextTo = nextTo
        self._cargaRobot = cargaRobot
        Robot._robots.append(self)

    #Metodos de busqueda
    def escuchar(self, casa):
        if self.getHealth() > 0:
            for hab in casa:
                if hab.getAlarma() is Ahorro.ENCENDIDO:
                    self._aware = True
                    self._goingTo = hab
                    break
                else:
                    self._aware = False

    def escanear(self):
        if self.getHealth() > 0:
            disponibles = []
            disponibles.append(self.getUbicacion().getNorte())
            disponibles.append(self.getUbicacion().getSur())
            disponibles.append(self.getUbicacion().getEste())
            disponibles.append(self.getUbicacion().getOeste())
            for Hab in disponibles:
                if (Hab is not None) and Hab.getLuces() is Ahorro.ENCENDIDO:
                    self._nextTo = True
                    self._goingTo = Hab
                    break
                else:
                    self._nextTo = False

    def apagarAlarma(self):
        if self.getHealth() > 0 and self.getUbicacion().getAlarma() is Ahorro.ENCENDIDO:
            self.getUbicacion().setAlarma(Ahorro.ACTIVADO)
    def buscar(self, casa):
        hab = None
        if self.getHealth() > 0:
            disponibles = []
            disponibles.append(self.getUbicacion().getNorte())
            disponibles.append(self.getUbicacion().getSur())
            disponibles.append(self.getUbicacion().getEste())
            disponibles.append(self.getUbicacion().getOeste())
            j = 0
            for d in disponibles:
                if d is None:
                    j += 1
            if (self._goingTo.getNorte() is not None) and self._goingTo.getNorte() in disponibles:
                hab = self._goingTo.getNorte()
            elif (self._goingTo.getSur() is not None) and self._goingTo.getSur() in disponibles:
                hab = self._goingTo.getSur()
            elif (self._goingTo.getEste() is not None) and self._goingTo.getEste() in disponibles:
                hab = self._goingTo.getEste()
            elif (self._goingTo.getOeste() is not None) and self._goingTo.getOeste() in disponibles:
                hab = self._goingTo.getOeste()
            elif j == 2:
                for d in disponibles:
                    if d is not None:
                        hab = d
            else:
                hab = casa[4]
        return hab
    
    #METODOS PELEAR
    def atacar(self, i):
        i.setHealth(i.getHealth()-Robot._ATTACK)

    def ataqueCargado(self, i):
        self._cargaRobot = False
        if Robot.lanzarDados(5) >= i.getArmor():
            self.atacar(i)
            self.atacar(i)
            self.atacar(i)
            return "El robot lanza un poderoso laser hacia ti!!"
        else:
            return "El robot lanza un poderoso laser hacia ti! Por suerte, logras esquivarlo"
    def turno(self, desicion, i):
        if desicion < 6:
            if Robot.lanzarDados(5) >= i.getArmor():
                self.atacar(i)
                return "El robot te acaba de asestar un golpe"
            else:
                return "Bloqueaste el golpe del robot"

        elif desicion == 6 or desicion == 7:
            self._cargaRobot = True
            return "El pecho del robot comienza a brillar con fuerza"
        elif desicion == 8 or desicion == 9:
            if Robot.lanzarDados(5) >= i.getArmor():
                i.stun(True)
                return "El robot te electrocutó, estarás aturdido por el siguiente turno"
            else:
                return "Bloqueaste el golpe del robot"
        elif desicion == 10:
            if i.getObjectInventory():
                dados = Robot.lanzarDados(len(i.getObjectInventory()))-1
                quitada = i.getObjectInventory().pop(dados)
                self.getUbicacion().getListaObjetos().append(quitada)
                
                return "El robot logra quitarte " + quitada.getName() + " y lo arroja al suelo."
            else:
                return "El robot intenta quitarte algo... pero aun no has recogido nada."

        else:
            return ""

    #METODOS GET
    @staticmethod
    def getATTACK():
        return Robot._ATTACK
    def isAware(self):
        return self._aware
    def isNextTo(self):
        return self._nextTo
    def isCargaRobot(self):
        return self._cargaRobot
    def getGoingTo(self):
        return self._goingTo

    #METODOS SET
    def setAware(self, b):
        self._aware = b

    def ayudaJarvis(self):
        if self._aware or self._nextTo:
            a = "sabe dónde estás."
        else:
            a = "no nota tu presencia... aún."
        if self.getHealth() > 0:
            return "El robot se encuentra en la habitacion " + str(self.getUbicacion().getNumero()) + ", tiene " + str(self.getHealth()) + " puntos de vida y " + a
        else:
            return "El robot ya ha sido destruido"

    def addHistorial(self):
        Robot.getHistorial().append("Robot se movio a la habitacion " + str(self.getUbicacion().getNumero()))
    def mover(self, hab):
        if self.getHealth() > 0:
            self.getUbicacion().setRobot(None)
            self.setUbicacion(hab)
            self.getUbicacion().setRobot(self)
            self.addHistorial()
        else:
            self.setUbicacion(None)

    def decidirDireccion(self):
        disponibles = []
        disponibles.append(self.getUbicacion().getNorte())
        disponibles.append(self.getUbicacion().getSur())
        disponibles.append(self.getUbicacion().getEste())
        disponibles.append(self.getUbicacion().getOeste())
        i = []
        for Hab in disponibles:
            if Hab is not None:
                i.append(Hab.getNumero())
        return i[Robot.lanzarDados(len(i))-1]

    def toString(self):

        m = "El Robot tiene... vida: " + self.getHealth() + " armadura: " + self.getArmor() + " velocidad: " + self.getSpeed() + " ataque: " + str(Robot._ATTACK)
        m += ", esta en la habitacion " + self.getUbicacion().getNumero()
        return m

